src/pkgcore/ebuild/repo_objs.py

```diff
@@ -170,6 +171,217 @@ def __init__(self, metadata_xml, manifest):
+        self.role = role
+        self.is_lead = is_lead
+
+    def __str__(self):
+        if self.name is not None:
+            res = f'{self.name} <{self.email}>'
+        else:
+            res = self.email
+        if self.role is not None:
+            return f'{res} ({self.role})'
+        return res
+
+
+class Subproject(object):
+    """Data on a subproject.
+
+    :type inherit_members: C{bool}
+    :ivar inherit_members: whether the parent project inherits members
+                           from this subproject.
+    """
+
+    __slots__ = ('_ref', 'inherit_members', '_projects_xml',
+                 '_cached_project')
+
+    def __init__(self, ref, projects_xml, inherit_members=None):
+        if ref is None:
+            raise ValueError('ref for subproject must not be null')
+        self._ref = ref
+        self.inherit_members = inherit_members
+        self._projects_xml = projects_xml
+        self._cached_project = None
+
+    @property
+    def _project(self):
+        if self._cached_project is None:
+            self._cached_project = self._projects_xml.projects[self._ref]
+        return self._cached_project
+
+    __getattr__ = klass.GetAttrProxy("_project")
+    # TODO: radhermit, why doesn't this work? because of __slots__?
+    __dir__ = klass.DirProxy("_project")
+
+
+class Project(object):
+    """Data on a single project.
+
+    :type email: C{unicode} object
+    :ivar email: email address.
+    :type name: C{unicode} object or C{None}
+    :ivar name: full name
+    :type url: C{unicode} object or C{None}
+    :ivar url: project website URI
+    :type description: C{unicode} object or C{None}
+    :ivar description: full project description.
+    :type members: C{tuple} of C{ProjectMember}
+    :ivar members: project members
+    :type subprojects: C{tuple} of C{Subprojects}
+    :ivar subprojects: subprojects
+    """
+
+    __slots__ = ('email', 'name', 'url', 'description', 'members',
+                 'subprojects')
+
+    def __init__(self, email, name=None, url=None, description=None,
+                 members=(), subprojects=()):
+        if email is None:
+            raise ValueError('email for project must not be null')
+        self.email = email
+        self.name = name
+        self.url = url
+        self.description = description
+        self.members = members
+        self.subprojects = subprojects
+
+    def __str__(self):
+        if self.name is not None:
+            res = f'{self.name} <{self.email}>'
+        else:
+            res = self.email
+        if self.url is not None:
+            return f'{res} ({self.url})'
+        return res
+
+    @property
+    def leads(self):
+        """Project lead(s), if any."""
+        return tuple(m for m in self.members if m.is_lead)
+
+    @property
+    def recursive_members(self):
+        """All project members, including members inherited from subprojects."""
+        # collect all subprojects from which to inherit, recursively
```

Comment:
Are the levels of projects always restricted to be at most three by policy? Currently this code appears to cap out at project -> subproject -> subproject which isn't quite recursive.

Reply:
You sure about that? I don't see any reason why the code wouldn't go on forever, as long as new projects are appended to `subprojects` list.

Reply:
Nope, looks like I'm incorrect. However I'm still wondering, how many levels is the deepest we currently get?

Reply:
I have no clue and I don't feel like counting ;-P.